ocaml/xapi/xapi_pool_update.ml

```diff
@@ -532,21 +532,42 @@ let resync_host ~__context ~host =
     |> List.iter (fun self -> destroy ~__context ~self)
   end
 
-let path_from_uri uri =
+let path_and_host_from_uri uri =
   (* remove any dodgy use of "." or ".." NB we don't prevent the use of symlinks *)
-  String.sub_to_end uri (String.length Constants.get_pool_update_download_uri)
-  |> Filename.concat !Xapi_globs.host_update_dir
-  |> Uri.pct_decode
-  |> Stdext.Unixext.resolve_dot_and_dotdot
+  let host_and_path = String.sub_to_end uri (String.length Constants.get_pool_update_download_uri) in
+  match String.split ~limit:2 '/' host_and_path with
+  | [host; untrusted_path] ->
+    let resolved_path = untrusted_path
+      |> Filename.concat !Xapi_globs.host_update_dir
+      |> Uri.pct_decode
+      |> Stdext.Unixext.resolve_dot_and_dotdot
+    in
+    (host,resolved_path)
+  | _ -> failwith "Expecting "
+
+let proxy_request req s host_uuid =
+  Server_helpers.exec_with_new_task "pool_update_proxy_request" (fun __context ->
+    let host = Db.Host.get_by_uuid ~__context ~uuid:host_uuid in
+    let ip = Db.Host.get_address ~__context ~self:host in
+    let transport = Xmlrpc_client.(SSL(SSL.make (), ip, 443)) in
+    Xmlrpc_client.with_transport transport (fun fd ->
+      Unixext.really_write_string fd (Http.Request.to_wire_string req);
+      Unixext.proxy s fd)
+  )
 
 let pool_update_download_handler (req: Request.t) s _ =
   debug "pool_update.pool_update_download_handler URL %s" req.Request.uri;
-  let filepath = path_from_uri req.Request.uri in
+  let localhost_uuid = Helpers.get_localhost_uuid () in
+  let (host_uuid, filepath) = path_and_host_from_uri req.Request.uri in
   debug "pool_update.pool_update_download_handler %s" filepath;
-  if not(String.startswith !Xapi_globs.host_update_dir filepath) || not (Sys.file_exists filepath) then begin
-    debug "Rejecting request for file: %s (outside of or not existed in directory %s)" filepath !Xapi_globs.host_update_dir;
-    Http_svr.response_forbidden ~req s
-  end else begin
-    Fileserver.response_file s filepath;
-    req.Request.close <- true
-  end
+  if host_uuid <> localhost_uuid then
+    proxy_request req s host_uuid
```

Comment:
I think the `host_uuid` comes from `Db.Host.get_uuid ~__context ~self:host`.
And for safety, we can check the uuid to be a valid host uuid here. @jonludlam

Reply:
When a XenServer slave sends it then it is valid, but this is an unauthenticated API, anyone can send anything here, and we should be robust against that. If we raise an exception when we detect the invalid host uuid that should be enough though.

Reply:
We already have it in `proxy_request`: `let host = Db.Host.get_by_uuid ~__context ~uuid:host_uuid in`

Reply:
Step further, is it possible to restrict the proxy only used for the request comes from localhost? Aka the client outside the localhost cannot use this proxy. It will be safer.

Reply:
Unfortunately not, since the external stunnel reconnects over localhost, hence xapi can't tell the difference between a call that really honestly originated from yum running on localhost and stunnel running on localhost.

In any case, what will happen is that you'll get an exception raised by the `Db.Host.get_by_uuid`. The only thing we need to do is verify that at least the fd is closed, and preferably we send a reasonable HTTP error before we close it.

Reply:
Hi, all. I have updated the code.
Trying to catch the exception for `Db.Host.get_by_uuid`, if invalid then returns a bad http request, finally close the fd.
@jonludlam

Reply:
There is already an exception handler at https://github.com/xapi-project/xen-api-libs-transitional/blob/master/http-svr/http_svr.ml#L457 and for this case, the code will send a HTTP 500 error and close the socket automatically.
IMO the `proxy_request` is fine, we don't need to handle the exception, just leave the work to upper layer.

Reply:
I agree with @krizex. And I think there has a very low probability to have a bad uuid here, the exception will handle by upper layer too. @robhoes  Needs your review and confirm. Thank you.